core/imds/src/main/java/software/amazon/awssdk/imds/internal/AsyncTokenCache.java

```diff
@@ -0,0 +1,138 @@
+ * A call to the {@link AsyncTokenCache#get()} method returns an already completed future if the cached token is still fresh.
+ * If the cached token was expired, the returned future will be completed once the refresh process has been
+ * completed.
+ * In the case where multiple call to <pre>get</pre> are made while the token is expired, all CompletableFuture returned
+ * will be completed once the single refresh process completes.
+ *
+ */
+@SdkInternalApi
+final class AsyncTokenCache implements Supplier<CompletableFuture<Token>> {
+
+    private static final Logger log = Logger.loggerFor(AsyncTokenCache.class);
+
+    /**
+     * The currently cached value.
+     * Only modified through synchronized block, under the refreshLock.
+     */
+    private volatile Token cachedToken;
+
+    /**
+     * The asynchronous operation that is used to refresh the token.
+     * The Supplier must not block the current thread and is responsible to start the process that will complete the future.
+     * A call the {@link AsyncTokenCache#get} method does not join or wait for the supplied future to finish, it only refreshes
+     * the token once it finishes.
+     */
+    private final Supplier<CompletableFuture<Token>> supplier;
+
+    /**
+     * The collection of future that are waiting for the refresh call to complete. If a call to {@link AsyncTokenCache#get()}
+     * is made while the token request is happening, a future will be added to this collection. All future in this collection
+     * are completed once the token request is done.
+     */
+    private Collection<CompletableFuture<Token>> waitingFutures;
+
+    /**
+     * Indicates if the token refresh request is currently running or not.
+     */
+    private final AtomicBoolean refreshRunning;
+
+    private final Object refreshLock = new Object();
+
+    AsyncTokenCache(Supplier<CompletableFuture<Token>> supplier) {
+        this.supplier = supplier;
+        this.waitingFutures = new ArrayList<>();
+        this.refreshRunning = new AtomicBoolean(false);
+    }
+
+    @Override
+    public CompletableFuture<Token> get() {
+        Token currentValue = cachedToken;
+        if (!needsRefresh(currentValue)) {
+            log.debug(() -> "IMDS Token is not expired");
+            return CompletableFuture.completedFuture(currentValue);
+        }
+        synchronized (refreshLock) {
+            // Make sure the value wasn't refreshed while we were waiting for the lock.
+            currentValue = cachedToken;
+            if (!needsRefresh(currentValue)) {
+                return CompletableFuture.completedFuture(currentValue);
+            }
+            CompletableFuture<Token> result = new CompletableFuture<>();
+            if (refreshRunning.get()) {
+                waitingFutures.add(result);
+            } else {
+                CompletableFuture<Token> tokenRequest = startRefresh(result);
+                CompletableFutureUtils.forwardExceptionTo(result, tokenRequest);
```

Comment:
We probably don't want this. A cancellation of the result future probably shouldn't stop the refresh that's happening in the background, because other calls might be waiting on that refresh.

Reply:
I see what you mean. I think it's fine to remove it because we will still cancel the request made with the token after, it's only the token refresh call that would be not be cancellable by the users.